concepts/type-theory-glossary.md

```diff
@@ -0,0 +1,396 @@
+that must be present on a type for it to satisfy the interface;
+then, any type that has those methods implements the interface.
+
+Both explicit and structural styles of interfaces have "open" membership --
+you can "compile" a program that has interfaces in it,
+then add more libraries when you "link" the program and execute it,
+and those libraries can add more implementors of the interfaces!
+
+The explicit and structural styles vary in other practical details.
+For example, explicit interfaces require referencing the interface they implement;
+this makes them significantly less flexible, and creates many dependency management requirements.
+By contrast, structural interfaces can knowingly conform to an interfaces contract
+while still not referencing it directly; this can be extremely useful in
+designing systems for independently upgradeable pieces.
+
+(There are other yet more detailed ways in which these styles vary, but
+since they're not really important in this document, we'll rest here.)
+
+
+Types in the Wild
+-----------------
+
+These different categories of types are present in most programming languages,
+but not all of them are present in every language (!),
+and they often go by different names in different programming communities.
+Here's a few (non-exhaustive) examples in a handful of languages:
+
+### enums in the wild
+
+In Java, the `enum` keyword maps directly the category of [enum](#enums) we've defined here.
+
+In Go, *there is no direct support for enums*.
+However, it's often implemented in practice by a typedef of an int, and a table of named constants.
+
+In Rust, there's an `enum` keyword, and it may be used to implement [enum](#enums) as we've defined it here.
+However, if you're coming from a Rust background, and trying to understand this document and compare to other languages,
+be cautious here: the Rust `enum` keyword *also* provides what we call [sum types](#sum-types).
+
+#### booleans
+
+In many languages, the concept of `boolean` can be understood as a particular instance of an enum --
+it's just an enum with two values!
+Sure, it may be implemented as 'primitive' in many languages... but clearly,
+for cardinality counting, we can see a boolean as exactly "|2|".
+
+### sum types in the wild
+
+In Java, *there is no direct support for sum types*.
+The nearest thing possible is to create an interface,
+and promise that only a known number of types will implement it.
+
+In Go, *there is no direct support for sum types*.
+The nearest thing possible is to create an interface with an unexported method,
+and make sure only a known number of types in the same package implement it.
+(See https://github.com/BurntSushi/go-sumtype for tooling to help support this;
+the readme also contains good documentation of the concept!)
+
+In Rust, the `enum` keyword (confusingly) also creates sum types.
+Rust's sum types are at least properly closed (unlike Java's and Go's).
+
+#### maybe types
```

Comment:
Do languages that support proper `null` (JavaScript, Java, not Go—`nil` is weird) have this in a basic form, "may be concrete or may not exist"?

Reply:
It's an H4, not an H3.  It's just an example of sum type.

I'm afraid expanding this section to list examples would be terribly dry: "Lang1 calls it Option.  Lang2 calls it Optional.  Lang3 calls it Maybe.  Lang4 calls it Option.  Lang5 calls it Either.", etc.

Yeah, I don't wanna talk about Go's typed nils.

Having a null is somewhat different and somewhat the same.  Null generally means a "+1", which makes it the same.  But when we talk about null, we generally talk about languages that allow null *anywhere*, which turns into a different beast, because it's like *every* Type *always* having a Maybe attached to it, rather than the Maybe being something you attach when appropriate.  Having null be an option *everywhere*, with no way to outlaw it, ends up being contrary to building cardinality-correct representations of data.

Reply:
... hm.  Maybe "nullable is sorta a sum type" deserves another H4 around here.

Better phrasing than "sorta" wanted, though.  So I'm probably going to leave this for "later".